keps/sig-node/00014-resource-api.md

```diff
@@ -0,0 +1,330 @@
+   // This will allow a single ResourceClass (e.g., “gpu”) to match multiple types of 
+   // resources (e.g., nvidia.com/gpu and amd.com/gpu) through general set selector.
+   Properties map[string]string
+}
+struct ComputeResourceStatus {
+   Capacity resource.Quantity
+}
+```
+
+ComputeResource objects are similar to PV objects on storage. It is tied to the physical resource that can have a wide range of vendor specific properties. Kubelet will create or update ComputeResource objects upon any resource availability or property change for node-level resources. Once a node is configured to support ComputeResource API and the underlying resource is exported as a ComputeResource, its quantity should NOT be included in the conventional NodeStatus Capacity/Allocatable fields to avoid resource multiple counting. ComputeResource API can be included into NodeStatus to facilitate resource introspection.
+For cluster level resources, a special controller or a scheduler extender can create a ComputeResource and dynamically bind that to a node during or after scheduling.
+
+### ResourceClass API
+ComputeResource mostly serves as an internal API to communicate resource states between different Kubernetes components, like Kubelet and the Scheduler. We will introduce ResourceClass as an additional abstraction layer for better portability and easy management, similar to why StorageClass was introduced to address similar requirements on storage. The scheduler will need to track both ResourceClass and ComputeResource objects to support resource requests expressed through ResourceClass names. 
+
+During the initial phase, we propose to start with the following ResourceClass API spec that defines the basic ResourceClass name to the underlying compute resource plus metadata matching. We can extend this API to include auto provisioning configurations and support group resources in the following phases of development. However, this document will mostly focus on the design to support initial phase requirements.
+
+```golang
+// +nonNamespaced=true
+// +genclient=true
+
+type ResourceClass struct {
+        metav1.TypeMeta
+        metav1.ObjectMeta
+        Spec ResourceClassSpec
+        // +optional
+}
+
+type ResourceClassSpec struct {
+        // defines general resource property matching constraints.
+        // e.g.: zone in { us-west1-b, us-west1-c }; type: k80
+        MetadataRequirements metav1.LabelSelector
+        // used to compare preference of two matching ResourceClasses
+        // The purpose to introduce this field is explained more later
+        Priority int
+}
+```
+
+YAML example 1:
+```yaml
+kind: ResourceClass
+metadata:
+  name: nvidia.high.mem
+spec:
+  labelSelector:
+    - matchExpressions:
+        - key: "Kind"
+          operator: "In"
+          values:
+            - "nvidia-gpu"
+        - key: "memory"
+          operator: "GtEq"
+          values:
+            - "30G"
+```
```

Comment:
Mind to add another yaml of how Pod reference `ResourceClass`?

Reply:
good idea. will do.